### ISSUE TITLE
Suggestion: Improve performance of .consumption template, only unzip and deserialize once

### ISSUE THREAD

**Issue by torronen** [dotnet/machinelearning-modelbuilder#1567]
**Is your feature request related to a problem? Please describe.**
Current .consumption file has low performance for multiple runs. This would seem to be cause in big part due to unzip and deserialization of the model. For an example, I might predict "suitability" score for a user for all movies he browse in a list. Each page may have 10-20 predictions. Currently .consumption file recreates the model object from the zip file every time

**Describe the solution you'd like**
Ideally, the .consumption file would be high performance multi-threaded version which does not require model deserialization for every call.

I would like to be able to edit the template file for .consumption to suit my specific needs.
If not possible,  the community could make pull requests with improvement requests it if it were in Github for a improved shared version.

**Possible ideas for solutions**
Moving the creation of model object to static constructor helps to unzip and deserialize once. The problem with this approach is parallel calls from multiple threads do not work.

To support multi-threading round robin could help which I am now using. The implementation is not good, suggestions for improvement are appreciated. 

```
        static List<PredictionEngine<ModelInput, ModelOutput>> robin = null;
        static int i = 0; // track current position in round robin
        const int max = 32; // how many predictionengines to create, i.e. max threads to support

        public Model()
        {
            MLContext mlContext = new MLContext();

            // Load model & create prediction engine
            ITransformer mlModel = mlContext.Model.Load(MLNetModelPath, out var modelInputSchema);
            robin = new List<PredictionEngine<ModelInput, ModelOutput>>();
            Parallel.For(0, max, (i) =>
            {
                var item = mlContext.Model.CreatePredictionEngine<ModelInput, ModelOutput>(mlModel);
                robin.Add(item);
            });
        }
```

**Alternatives**
Batch predictions method which could be called with a group of observations, but this does not solve all cases. For example, gaming may need rapid single predictions depending on the actions of the player.

**Comment by beccamc** [comment]
@briacht for visibility

**Comment by torronen** [comment]
It may need to be confirmed PredictionEngine has any static properties or otherwise is not thread-safe even with multiple instance.
The error https://github.com/dotnet/machinelearning-modelbuilder/issues/1610 occurs only simultaneous calls from multiple threads (to above round robin of multiple instances.

**Comment by torronen** [comment]
Proper way seems PredictionEnginePool https://docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/serve-model-web-api-ml-net#register-predictionenginepool-for-use-in-the-application

**Comment by briacht** [comment]
Hey @torronen! @LittleLittleCloud is working on improving consumption.cs / Predict function perf. Maybe he can share progress on that so far.

We definitely want to update it so that loading the model and creating the prediction engine only has to happen once.

PredictionEnginePool is what is recommended for web apps. We have talked about adding another method in consumption.cs, something like PredictScalable() which implements with the PredictionEnginePool. Non-web apps (desktop, console, etc.) can just use PredictionEngine

And then of course there's the Transform() method for batch predictions which could be another method for consuming a model

**Comment by torronen** [comment]
How about adding an interface to the template, such as : IMLModel?

I am running multiple models with same data, (maybe `Execute(IModel model, string[][] raw)` or  `Execute<T,Y>(T rawData, Y model) `) so I need some interface or parent calass. For now, I will add the interface declaration manually so not an issue for me, but might one improvement to consider (?)

**Comment by torronen** [comment]
Namespaces may be worth consideration, although not really a problem. My project names have often been "Kwork.ML. ..." with namespace Kwork.ML.....  Model Builder makes it Kwork_ML_. 

Keeping the dots from project name as the namespace might be better as it's in the .NET naming guidelines for namespaces https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-namespaces

**Comment by luisquintanilla** [comment]
Thanks for the suggestion. Looks like this issue was addressed by #1720.
